Init only proposal document (#3367)
dotnet/csharplang: proposals/init.md

These mechanisms are effective at allowing the construction of immutable data
but they do so by adding cost to the boiler plate code of types and opting
such types out of features like object and collection initializers. This means
developers must choose between easy of use and immutability.

Simple immutable object like `Point` requires twice as much boiler plate code
to support construction as it does to declare the type. The bigger the type 
the bigger the cost of this boiler plate:

```cs
struct Point
{
    public int X { get; }
    public int Y { get; }

    public Point(int X, int Y)
    {
        this.X = x;
        this.Y = y;
    }
}
```

The `init` modifier makes immutable objects more flexible by allowing the
caller to mutate the members during the act of construction. That means the 
object can participate in object initialzers and thus removes the need for 
all boiler plate code in the type. The `Point` type is now simply:

```cs
struct Point
{
    public int X { get; init; }
    public int Y { get; init; }
}
```

The consumer can then use object initializers to create the object

```cs
var p = new Point() { X = 42, Y = 13 };
```

## Detailed Design

### init members

> An init only property is declared by using the `init` accessor in place of the 

Author: Once I commit your edits I'll do another pass to make this more consistent.

`set` accessor:

```cs
class Student
{
    public string FirstName { get; init; }
    public string LastName { get; init; }
}
```

An instance property containing an `init` accessor is considered settable in
the following circumstances:

- During an object initializer

> - Inside an instance constructor of the containing or derived type

Author: Don't want to get into validators here since they don't have a concrete design. Think it will be more appropriate for validators to call out the expanded places `init` accessors can be called just as the init only proposal calls out the expanded places `readonly` fields can be mutated.

in the open construction phase. Hence an `init` accessor is allowed to take 
the following actions:
1. Call other `init` accessors available through `this`
1. Assign `readonly` fields declared on the same type

```cs
class Complex
{
    readonly int Field1;
    int Field2;
    int Prop1 { get; init ; }
    int Prop2
    {
        get => 42;
        set
        {
            Field1 = 13; // okay
            Field2 = 13; // okay
            Prop1 = 13; // okay
        }
    }
}
```

The ability to assign fields from an `init` accessor is limited to fields
declared on the same type as the accessor. It cannot be used to assign fields
in a base type. This rule ensures that type authors remain in control over the
mutability behavior of their type. Developers who do not wish to utilize 
`init` cannot be impacted from other types choosing to do so:

```cs
class Base
{
    internal readonly int Field;
    internal int Property
    {
        get => Field;
        init => Field = value; // Okay
    }
}

class Derived : Base
{
    internal readonly int DerivedField;
    internal int DerivedProperty
    {
        get => DerivedField;
        init
        {

>             DerivedField = 42; // Okay

Reviewer: Consider assigning `Property` in this one as an example of what would be ok.

            Field = 13; // Error Field is readonly
        }
    }

    public Derived()
    {
        Property = 42; // Okay 
        Field = 13; // Error Field is readonly
    }
}
```

> When `init` is used in a virtual property then all the overrides must also

Reviewer: Consider adding an example of this.

be marked as `init`. Likewise it is not possible to override a simple 
`set` with `init`.

An `interface` declaration can also particpate in `init` style initalization 
via the following pattern:

```cs
interface IPerson
{
    string Name { get; init; }
}

class Init
{
    void M<T>() where T : IPerson, new()
    {
        var local = new T()
        {
            Name = "Jared"
        };
        local.Name = "Jraed"; // Error
    }
}
```

Restrictions of this feature:

> - The `init` accessor can only be used instance properties

Reviewer: ```suggestion
- The `init` accessor can only be used on instance properties
```

- A property cannot contain both an `init` and `set` accessor
- All overrides of a property must have `init` if the base had `init`. This rule

> also applies to interface implementation.

Reviewer: Interface implementations today allow you to only specify `get`, but add a `set` in the implementation. I think it's work allowing the same thing for specifying `init` in the interface but allowing `set` in the implementation.

Author: I can definitely get the motivation for allowing a class to implement a `get` only property with `get; init;`. That follows from `set` and seems pretty reasonable. I'm less certain about allowing a `get; init;` property be implemented with `get; set;`. It would require us to generate an explicit implementation of the interface under the hood (modreq would break the signature match). That makes me hesitate. I think I'd need to see some user feedback to want to take that on.

### Metadata encoding 
Property `init` accessors will be emitted as a standard `set` accessor with
the return type marked with a modreq of `IsInitOnly`. This is a new type
which will have the following definition:

```cs
namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.All)]

>     public sealed class InitOnlyAttribute : Attribute

Reviewer: Take updates from https://github.com/dotnet/runtime/issues/34978#issuecomment-614845405.

Author: Huh ... I changed this in my machine. Must not have saved before submitting.

    {

    }
}
```

> The compiler will match the attribute by full name. There is no requirement

Reviewer: ```suggestion
The compiler will match the type by full name. There is no requirement
```

> that it appear in the core library. In the case there are mulitple attributes

Reviewer: ```suggestion
that it appear in the core library. If there are multiple attributes
```

Author: I also need to remove the use of attributes in this sentence

by this name available then the compiler will pick the one defined in the core 

> library should one exist.

Reviewer: And if one does not?

Author: Overall I think we should follow the same rules we have for looking up the `Extensions` attribute when you define an extension method. That has a lot of the same similarities here.

## Questions

### Breaking changes
One of the main pivot points in how this feature is encoded will come down to
the following question: 

> Is it a binary breaking change to remove `init` from a `set`?

> Removing `init`, and thus making a field or property fully writable is never

Reviewer: ```suggestion
Replacing `init` with `set`, and thus making a field or property fully writable is never
```

Reviewer: And everywhere you say `remove` below.

> a source breaking change. For fields it is never a binary breaking change 

Reviewer: Consider removing `fields`, it's not in the actual proposal.

either. Additionally removing it from a field is never a binary breaking 
change. The only behavior up in question is whether or not this remains 
true for a property. 

If we want to make the removal of `init` from a property a compatible change

> then it will force our hand on the modreq vs. attributes decision below. If 
> one the other hand this is seen as a non-interesting scenario then this will 

Reviewer: ```suggestion
then it will force our hand on the modreq vs. attributes decision below. If, 
on the other hand, this is seen as a non-interesting scenario then this will 
```

**Resolution**
This scenario is not seen as compelling by LDM.

### Modreqs vs. attributes
The emit strategy for `init` property accessors must choose between using 
attributes or modreqs when emitting during metadata. These have different 
trade offs that need to be considered.

Annotating a property set accessor with a modreq declaration means CLI compliant
compilers will ignore the accessor unless it understands the modreq. That means
only compilers aware of `init` will read the member. Compilers unaware of 
`init` will ignore the `set` member and hence will not accidentally treat the
property as read / write. 

The downside of modreq is `init` becomes a part of the binary signature of 
the `set` accessor. Adding or removing `init` will break binary compatbility 
of the application.

Using attributes to annotate the `set` accessor means that only compilers which
understand the attribute will know to limit access to it. A compiler unaware 
of `init` will see it as a simple read / write property and allow access.

This would seemingly mean this decision is a choice between extra safety at 
the expense of binary compatibility. Digging in a bit the extra safety is not

> exactly what it seems. It will not for instance protect against the following
> circumstances:

Reviewer: ```suggestion
exactly what it seems. It will not protect against the following circumstances:
```

1. Reflection over `public` members
1. The use of `dynamic` 
1. Compilers that don't recognize modreqs

> It should also be considered that when we complete the IL verification rules 

Reviewer: ```suggestion
It should also be considered that, when we complete the IL verification rules 
```

> for .NET 5, `init` will be one of those rules. that means extra enforcement 

Reviewer: ```suggestion
for .NET 5, `init` will be one of those rules. That means extra enforcement 
```

will be gained from simply verifying compilers emitting verifiable IL.

The primary languages for .NET (C#, F# and VB) will all be updated to 
recognize these `init` accessors. Hence the only realistic scenario here is 

> when a C# 9 compiler emits `init` properties and they are seen by a C# 8 

Reviewer: Or an older VB compiler

compiler. That is the trade off to consider and weigh against binary 
compatibility.

Note: this discussion applies to the accessor members only, not to fields. There
is no value to be gained by using a modreq on a field. The `init` feature for 

> fields is a relaxation of an existing rule. All existing compilers already 

Reviewer: We're also not doing the feature.

Author: I want to leave the field discussion in because I think it's important for the entirety of the argument here. The argument is meant to be more general. I antincipate referring to it again when we discuss modreqs for future features. I will add a mention that fields aren't being done though.

support `readonly` and hence an attribute serves fine as a way to alert them
that write access can be extended in certain circumstances.

**Resolution**
The feature will use a modreq to encode the property `init` setter. The
compelling factors were (in no particular order):

> 1. Desire to discourage older compilers from violating `init` semantics
> 1. Desire to make adding or removing `init` in a `virtual` declaratio or 

Reviewer: ```suggestion
* Desire to discourage older compilers from violating `init` semantics
* Desire to make adding or removing `init` in a `virtual` declaratio or 
```

Author: Have a typo on `declaratio` I need to fix here.

`interface` both a source and binary breaking change.

Given there was also no significant support for removing `init` to be a 
binary compatible change it made the choice of using modreq straight forward.

### init vs. initonly
Syntax debate time.

**Resolution**
There was no other syntax which was overwelming favored in LDM. For the time 
being we will be moving forward with the initial design laid out in this 
document. This may change as the feature progresses and LDM members have time

> to digest the design here.

Reviewer: Consider talking about `init set` vs `init`, and us being open to making `init` a shorthand for `init set` in the future if we decide to make `init` a modifier for functions.

1. It requires carrying additional metadata around to denote the members
which are required to be initialized by the caller.

Further if we believe there is value here in the overall scenario of forcing
object creators to be warned / error'd about specific fields then this 
likely makes sense as a general feature. There is no reason it should be 
limited to just `init` members.

**Resolution**
There will be no warning on consumption of `init` fields and properties.

LDM wants to have a broader discussion on the idea of required fields and
properties. That may cause us to come back and reconsider our position on
`init` members and validation.

### Allow init as a type modifier
In the same way the `readonly` modifier can be applied to a `struct` to 
automatically declare all fields as `readonly`, the `init` only modifier can
be declared on a `struct` or `class` to automatically mark all fields as `init`.
This means the following two type declarations are equivalent:

```cs
struct Point
{
    public init int X;
    public init int Y;
}

// vs. 

init struct Point
{
    public int X;
    public int Y;
}
```

**Resolution** 

> This feature is too *cute* here and conflicts with the `readonly struct` 

Reviewer: Consider removing the `too *cute*` part.

feature on which it is based. The `readonly struct` feature is simple in that 
it applies `readonly` to all members: fields, methods, etc ... The 

> `init struct` feature would only apply to fields. This actually ends up making

Reviewer: ```suggestion
`init struct` feature would only apply to properties. This actually ends up making
```

it confusing for users. 

Given that `init` is only valid on certain aspects of a type we rejected the 
idea of having it as a type modifier.

> ## Allow init as a field modifier

Reviewer: Consider moving this before the previous section, given that the decision was based on `init fields` not being a thing.

In the same way `init` can server as a property accessor it could also serve as
a designation on fields to give them similar behaviors as `init` properties.
That would allow for the field to be assigned before construction was complete
by the type, derived types or object initializers.

```cs
class Student
{
    public init string FirstName;
    public init string LastName;
}

var s = new Student()
{
    FirstName = "Jarde",
    LastName = "Parsons",
}

s.FirstName = "Jared"; // Error FirstName is readonly
```

In metadata these fields would be marked in the same way as `readonly` fields 
but with an additional attribute or modreq to indicate they are `init` style
fields. 

**Resolution**
LDM agrees this proposal is sound but overal the scenario felt disjoint from 
properties. The decision was to proceed with properties only for now but this 
will be reconsidered if there is significant customer feedback that justifies

> the scenario.

Reviewer: Consider also mentioning that this discussion was when we decided to make `readonly` fields settable in initers on the type.

## Considerations

### Compatibility
The `init` feature is designed to be compatible with existing data types. 
Specifically it is meant to be a completely additive change for data which is
`readonly` today but desires more flexible object creation semantics.

For example consider the following type:

```cs
class Name
{
    public readonly string First;
    public readonly string Last;

    public Name(string first, string last)
    {
        First = first;
        Last = last;
    }
}
```

> It is not a breaking change to use `init` in place of `readonly` here:

Reviewer: True for structs as well?

Author: This entire section is too field centric. Need to pretty much completely overhaul this.

```cs
class Name
{
    public init string First;
    public init string Last;

    public Name(string first, string last)
    {
        First = first;
        Last = last;
    }
}
```

> The same applies for changing a `get` only property to have an `init set` 

Reviewer: ```suggestion
The same applies for changing a `get` only property to have an `init` 
```

accessor.

### IL verification

> When .NET Core decides to re-implement IL verify the rules will need to be 

Reviewer: ```suggestion
When .NET Core decides to re-implement IL verification the rules will need to be 
```

adjusted to account for `init` members. This will need to be included in the 
rule changes for non-mutating acess to `readonly` data.

## init members
The `init` modifier could be extended to apply to all instance members. This 
would generalize the concept of `init` during object construction and allow
types to declare helper methods that could partipate in the construction 
process to initialize `init` fields and properties.

> Such members would have all the restricions that an `init set` accessor does

Reviewer: ```suggestion
Such members would have all the restricions that an `init` accessor does
```

```

The rules around when `init` accessors are settable extend across type
hierarchies. If the member is accessible and the object is known to be in the
construction phase then the member is settable. That specifically allows for
the following:

```cs
class Base
{
    public bool Value { get; init; }
}

class Derived : Base
{
    Derived()
    {
        // Not allowed with get only properties but allowed with init
        Value = true;
    }
}

class Consumption
{
    void Example()
    {
        var d = new Derived() { Value = true; };
    }
}

```

At the point a `init` accessor is invoked the instance is known to be 
in the open construction phase. Hence an `init` accessor is allowed to take 
the following actions in addition to what a normal `set` accessor can do:

1. Call other `init` accessors available through `this` or `base`
1. Assign `readonly` fields declared on the same type

```cs
class Complex
{
    readonly int Field1;
    int Field2;
    int Prop1 { get; init ; }
    int Prop2
    {
        get => 42;
        init
        {
            Field1 = 13; // okay
            Field2 = 13; // okay
            Prop1 = 13; // okay

>         }

Reviewer 2: Why do you want to add that, and what would be the advantage? To me, this is duplicating the job of a constructor, and I think this will only create less readable code. Maybe there is a use-case that I missed; if so, feel free to enlighten me.

Reviewer: I believe the motivation section of this proposal adequately outlines our reasoning: we want to support nominal, immutable construction patterns which we do not do today.

Reviewer 2: > I believe the motivation section of this proposal adequately outlines our reasoning: we want to support nominal, immutable construction patterns which we do not do today.

Yes, I get the intent @333fred , which is a really good idea; less boilerplate is always welcome. I still remember the day you added auto-implemented properties; what a day!

But what is the difference between the following two blocks of code?

```csharp
init
{
    Field1 = 13; // okay
    Field2 = 13; // okay
    Prop1 = 13; // okay
}
```

```csharp
public SomeCtor()
{
    Field1 = 13; // okay
    Field2 = 13; // okay
    Prop1 = 13; // okay
}
```

Then what is the added value of that new way to initialize fields/props? (for C# users)

An intend is one thing, but does that help solve a problem or does it only create noise and possible confusion? Think about it, people will have one more place to put initialization logic in, does someone need that? If yes, who does? What use-case does this solve that was unsolvable before or that required more boilerplate?

Reviewer: Think about optional parameters in a constructor today: you have a location that can default them if not provided, or throw if a combination of provided parameters is invalid. This adds the same logic for properties that are inited by the caller, but then immutable afterwards.

Reviewer 2: > Think about optional parameters in a constructor today: you have a location that can default them if not provided, or throw if a combination of provided parameters is invalid. This adds the same logic for properties that are inited by the caller, but then immutable afterwards.

When I think about that I see types that could require some kind of possible value/props/null/not null combo but without any way to know it in advance from the API of that class/struct. People can be creative 😉 

Then what about I add another property that initializes `init` and `readonly` props/fields as well?

```cs
class Complex
{
    readonly int Field1;
    int Field2;
    int Prop1 { get; init ; }

    int Prop2
    {
        get => 42;
        init
        {
            Field1 = 13; // okay
            Field2 = 13; // okay
            Prop1 = 13; // okay
        }
    }

    int Prop3
    {
        get => 123;
        init
        {
            Field1 = 3;
            Field2 = 3;
            Prop1 = 3;
        }
    }
}
```

In what order are those going to be initialized?

Wouldn't it be clearer to do the following?

```cs
class Complex
{
    readonly int Field1;
    int Field2;
    int Prop1 { get; init ; }
    int Prop2 { get; init; }
    int Prop3 { get; init; }

    public Complex()
    {
        Field1 = 13;
        Field2 = 13;
        Prop1 = 13;
        Prop2 = 42;
        Prop3 = 123;
    }
}
```

Reviewer: You shouldn't necessarily take syntax examples from a proposal that merely illustrate what things are legal where as a good API design. More than likely, most `init` properties would initialize one, and only one, readonly field. You can still write your example today, using `set` instead of `init`, and it provides a markedly worse experience if you wanted it to actually be an immutable object.

As to ordering, that's the thing here: this is about _nominal_ object creation. Nominal creation is unordered, that's the whole point. If you want to control the order in which properties are set, then you need to use the existing tools we have today in positional creation (namely, standard constructor arguments).

Imagine you had an example like this:

```cs
public class PathObject
{
    string ActualPath { get; init; }
    bool SupportsLongPath { get; init; }
    init
    {
        if (ActualPath == null || !supportsLongPath && ActualPath.Length >= 256)
        {
            throw new ArgumentException(nameof(ActualPath), $"Long path support not enabled, but path is {ActualPath.Length}!");
        }
    }
}
```

Imagine if you had to write that without `init` support? You literally can't do it. There's no way today to ensure that `ActualPath` was initialized by the caller. Further, for validation of the length of `ActualPath`, you can't do that either: you need to ensure that `SupportsLongPath` is set _before_ attempting to validate the path length. The only way to handle that correctly would be to use a constructor, which guarantees ordering. However, that brings burdens of its own. Let's say that this is a public class in a library that you publish publicly. You decide that you want to add support for hidden path object creation. But there's a problem: you've already published this class. If you want to avoid breaking your customers, you're going to have to maintain this 2-parameter constructor forever more, because removing it would break existing customer binaries. In the `init` world, this pain point is gone. You add a new init-only property, validate as necessary in the `init` body, and no one who compiled against a previous version of your library has to care about this new optional parameter that you've added. For a very real example of this type of pain, look at https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs#L787-L913. We cannot remove these constructors, as existing binaries depend on them, even the constructor marked as `Bad - DO NOT USE`. Those constructors will never be able to go away, unfortunately, but at least we won't have to add more of them.

Reviewer 2: Good point, I could write my messy code in a `set`, so `init` is similar in that way (well, its an "initializable `readonly` property setter" after all).

That said, I did not see the "init constructor-like" that you used in your last sample anywhere in the proposal, maybe you should add that as it seems that will be needed.

If the initialization is unordered, how can you be sure that your constructor-like-init code block will be hit **after** the init setter of the other properties?

More on that, to make the usage clearer for users, maybe `afterinit` would be easier to understand for non-compiler/csharplang-guru? It would also point people in the direction of centralizing initialization logic instead of splitting it everywhere. Maybe a `beforeinit` could benefit centralizing initialization while `afterinit` benefits validation?

Or to make it less "magic" and more discoverable, what about an `IInitializable` interface that adds a `void AfterInit();` and a `void BeforeInit()` methods usable during the initialization process? Same idea here, you need additional logic, you implement the interface (or create the `afterinit` block) and code your initialization logic there; otherwise, you don't pollute your code with it.

_As I understand it, you seem to want to help people get into C#, so making usage and understanding of new (and existing) features as easy as possible for everyone would probably help that. Just for fun, ask people in what order are executed auto-property initializers and constructors to see how unclear it can be. With `init` you will be adding another layer of "confusion" over that._

As for breaking changes, a new major version should allow you to clean stuff up and introduce breaking changes. Everyone makes mistakes but having to carry all of the mistakes of everyone forever ain't helping anyone 😉

Reviewer: > That said, I did not see the "init constructor-like" that you used in your last sample anywhere in the proposal, maybe you should add that as it seems that will be needed.

It'll be coming in a separate proposal. It'll all part of a greater feature unity we're aiming for as part of records.

> If the initialization is unordered, how can you be sure that your constructor-like-init code block will be hit after the init setter of the other properties?

The compiler will be required to generate a call to the validator (straw-man name we're using for now) after an object initializer is run, so ordering won't matter.

Don't take the syntax I used as "How it will look exactly". A future proposal with more validator design will be coming, and we haven't talked much about the specific syntax beyond "We like the general idea".

Reviewer 2: Good to know, thanks, I'll try to follow that up then.

binary compatible change it made the choice of using modreq straight forward.

### init vs. initonly
There were three syntax forms which got significant consideration during our
LDM meeting:

```cs
// 1. Use init 
int Option1 { get; init; }
// 2. Use init set
int Option2 { get; init set; }
// 3. Use initonly
int Option3 { get; initonly; }
```

**Resolution**
There was no syntax which was overwhelmingly favored in LDM.

One point which got significant attention was how the choice of syntax would
impact our ability to do `init` members as a general feature in the future.
Choosing option 1 would mean that it would be difficult to define a property
which had a `init` style `get` method in the future. Eventually it was decided
that if we decided to go forward with general `init` members in future we could
allow `init` to be a modifier in the property accessor list as well as a short
hand for `init set`. Essentially the following two declarations would be
identical.

```cs
int Property1 { get; init; }
int Property1 { get; init set; }
```

The decision was made to move forward with `init` as a standalone accessor in
the property accessor list.

> ### Warn on failed init

Reviewer 3: Any plans to add some option or another keyword that would force property to be set in initializer or constructor?

Reviewer 4: We've touched on that topic in recent discussions, but haven't made any plans for that yet.

> Init Only Members

Reviewer 4: ```suggestion
Init-Only Members
```
```suggestion
Init-Only Setters
```

> Init Only Members

Author: Discussed to hyphen or not to hyphen earlier with @333fred  and decided against it.

=====

## Summary

> This proposal adds the concept of init only properties to C#. These properties

Reviewer 4: ```suggestion
This proposal adds the concept of init-only setters in properties and indexers to C#. These properties and indexers
```

Reviewer: Ajred specifically went through and settled on `init only` throughout the document.

Reviewer 4: Missed that. Seems wrong though, but the hyphen is a nit, so I'm fine with whatever Jared picks.

Author: > Not sure how to avoid repeating "properties or indexers" in the whole document.

I think it's okay to just refer properties in the document and call out that it also works for indexers in the details section.

Reviewer 4: @jaredpar You may have missed some of my comments. The spec still doesn't mention indexers. I'll start a PR to correct some things.

---
In reply to: [411695374](https://github.com/dotnet/csharplang/pull/3367#discussion_r411695374) [](ancestors = 411695374)

Author: Didn't miss that comment. I added a mention of that + an explicit example using indexers to drive home it was intended to be supported. What I apparently missed was saving and pushing that to the doc 😦 

Saw you already updated the doc though to mention they're included so I think we're fine.

can be set at the point of object creation but become effectively `get` only 
once object creation has completed. This allows for a much more flexible
immutable model in C#. 

## Motivation
The underlying mechanisms for building immutable data in C# haven't changed
since 1.0. They remain:

1. Declaring fields as `readonly`.

> 1. Declaring properties that contain only a `get` accessor.

Reviewer 5: ```suggestion
2. Declaring properties that contain only a `get` accessor.
```

Reviewer 6: [As per above](https://github.com/dotnet/csharplang/pull/3367#discussion_r411008323) this is intentional.

These mechanisms are effective at allowing the construction of immutable data

> but they do so by adding cost to the boiler plate code of types and opting

Reviewer 4: ```suggestion
but they do so by adding cost to the boilerplate code of types and opting
```

such types out of features like object and collection initializers. This means
developers must choose between ease of use and immutability.

A simple immutable object like `Point` requires twice as much boiler plate code
to support construction as it does to declare the type. The bigger the type 
the bigger the cost of this boiler plate:

```cs
struct Point
{
    public int X { get; }
    public int Y { get; }

>     public Point(int X, int Y)

Reviewer 7: Parameters should be small "x" and "y".

Reviewer 7: ```suggestion
    public Point(int x, int y)
```

    {
        this.X = x;
        this.Y = y;
    }
}
```

> The `init` modifier makes immutable objects more flexible by allowing the

Reviewer 4: ```suggestion
The `init` accessor makes immutable objects more flexible by allowing the
```

caller to mutate the members during the act of construction. That means the

> object's immutable properties can participate in object initialzers and thus

Reviewer 5: ```suggestion
object's immutable properties can participate in object initializers and thus
```

> removes the need for all constructor boiler plate in the type. The `Point`

Reviewer 4: ```suggestion
removes the need for all constructor boilerplate in the type. The `Point`
```

type is now simply:

```cs
struct Point
{
    public int X { get; init; }
    public int Y { get; init; }
}
```

The consumer can then use object initializers to create the object

```cs
var p = new Point() { X = 42, Y = 13 };
```

## Detailed Design

### init members
An init only property is declared by using the `init` accessor in place of the 
`set` accessor:

```cs
class Student
{
    public string FirstName { get; init; }
    public string LastName { get; init; }
}
```

> An instance property containing an `init` accessor is considered settable in

Reviewer 4: ```suggestion
An instance property containing an `init` accessor is considered settable on `this` or `base` in
```

the following circumstances:

- During an object initializer

> - Inside an instance constructor of the containing or derived type

Reviewer 4: ```suggestion
- Inside an instance constructor of the containing or derived type, on `this` or `base`
```

> - Inside the `init` accessor of any property on `this`

Reviewer 4: ```suggestion
- Inside the `init` accessor of any property, on `this` or `base`
```

The times above in which the `init` accessors are settable are collectively
referred to in this document as the construction phase of the object.

This means the `Student` class can be used in the following ways:

```cs
var s = new Student()
{
    FirstName = "Jared",

>     LastName = "Parosns",

Reviewer 8: Typo here

};
s.LastName = "Parsons"; // Error: LastName is not settable
```

The rules around when `init` accessors are settable extend across type
hierarchies. If the member is accessible and the object is known to be in the
construction phase then the member is settable. That specifically allows for
the following:

```cs
class Base
{
    public bool Value { get; init; }
}

class Derived : Base
{
    Derived()
    {
        // Not allowed with get only properties but allowed with init
        Value = true;
    }
}

class Consumption
{
    void Example()
    {
        var d = new Derived() { Value = true; };
    }
}

```

At the point a `init` accessor is invoked the instance is known to be 
in the open construction phase. Hence an `init` accessor is allowed to take 

> the following actions in addition to what a normal `set` accessor cand do:

Reviewer 9: can do?

Author: Fixed

> 1. Call other `init` accessors available through `this`

Reviewer 4: ```suggestion
1. Call other `init` accessors available through `this` or `base`
```

1. Assign `readonly` fields declared on the same type

```cs
class Complex
{
    readonly int Field1;
    int Field2;

>     int Prop1 { get; init ; }

Reviewer 10: whitespace after init

Author: Thanks. Fixed.

    int Prop2
    {
        get => 42;

>         set

Reviewer 6: This should be `init` right? Otherwise I don't get why those assignments would be valid.

>         set

Reviewer 9: Shouldn’t this be `init`?

        {
            Field1 = 13; // okay
            Field2 = 13; // okay
            Prop1 = 13; // okay
        }
    }
}
```

The ability to assign `readonly` fields from an `init` accessor is limited to 
those fields declared on the same type as the accessor. It cannot be used to 
assign `readonly` fields in a base type. This rule ensures that type authors
remain in control over the mutability behavior of their type. Developers who do
not wish to utilize `init` cannot be impacted from other types choosing to
do so:

```cs
class Base
{

>     internal readonly int Field;

Reviewer 6: Maybe make the field lower-case?

Author: For design docs I usually leave all members as PascalCased. Feel like it makes the document easier to read.

Reviewer 6: Alright, I thought the exact opposite while reading this but I’m fine with it if this is like a design docs convention 😀

    internal int Property
    {
        get => Field;
        init => Field = value; // Okay
    }

    internal int OtherProperty { get; init; }
}

class Derived : Base
{
    internal readonly int DerivedField;
    internal int DerivedProperty
    {
        get => DerivedField;
        init
        {
            DerivedField = 42;  // Okay
            Property = 0;       // Okay
            Field = 13;         // Error Field is readonly
        }
    }

    public Derived()
    {
        Property = 42;  // Okay 
        Field = 13;     // Error Field is readonly
    }
}
```

When `init` is used in a virtual property then all the overrides must also
be marked as `init`. Likewise it is not possible to override a simple 
`set` with `init`.

```cs
class Base
{
    public virtual int Property { get; init; }
}

class C1 : Base
{
    public override int Property { get; init; }
}

class C2 : Base
{
    // Error: Property must have init to override Base.Property
    public override int Property { get; set; }
}
```

> An `interface` declaration can also particpate in `init` style initalization 

Reviewer 6: Typo: “participate”

Author: Thanks! Noticed you added a lot of typo style suggestions. If you want you can always use the "Insert Suggestion" for these types of items. The upsides are I can just click to accept your suggestion and your GitHub alias ends up in the document history. 

The click to accept is great because it's much harder for me to insert a different typo when fixing the first :smile:

![image](https://user-images.githubusercontent.com/146967/79701718-30741880-8254-11ea-9803-b7606d4cd554.png)

Reviewer 6: Yeah, sorry about that. I would have made edit suggestions if GitHub was allowing me to do so from the mobile website – I was sitting on the couch, reviewing this from my phone 😅🙈
If I knew before that I would have ended up with this many points, I maybe would have bothered to stand up and continue on my computer.
Sorry! 😅

Author: > Yeah, sorry about that. I would have made edit suggestions if GitHub was allowing me to do so from the mobile website – I was sitting on the couch, reviewing this from my phone

Not a problem. Mostly I wanted to make sure you were aware of the feature so you could get credit in the document history if that was interesting to you.

via the following pattern:

```cs
interface IPerson
{
    string Name { get; init; }
}

class Init
{
    void M<T>() where T : IPerson, new()
    {
        var local = new T()
        {
            Name = "Jared"
        };
        local.Name = "Jraed"; // Error
    }
}
```

Restrictions of this feature:
- The `init` accessor can only be used on instance properties
- A property cannot contain both an `init` and `set` accessor
- All overrides of a property must have `init` if the base had `init`. This rule
also applies to interface implementation.

### Metadata encoding 
Property `init` accessors will be emitted as a standard `set` accessor with
the return type marked with a modreq of `IsInitOnly`. This is a new type
which will have the following definition:

```cs
namespace System.Runtime.CompilerServices
{
    public sealed class IsInitOnly
    {

> 

Reviewer 4: ```suggestion
```

    }
}
```

The compiler will match the type by full name. There is no requirement

> that it appear in the core library. If there are mulitple types by this name

Reviewer 11: ```suggestion
that it appear in the core library. If there are multiple types by this name
```

1. The one defined in the project being compiled
1. The one defined in corelib

If neither of these exist then a type ambiguity error will be issued.

The design for `IsInitOnly` is futher covered in [this issue](https://github.com/dotnet/runtime/issues/34978)

## Questions

### Breaking changes
One of the main pivot points in how this feature is encoded will come down to
the following question: 

> Is it a binary breaking change to replace `init` with `set`?

Replacing `init` with `set` and thus making a property fully writable is never
a source breaking change on a non-virtual property. It simply expands the set
of scenarios where the property can be written. The only behavior in question is
whether or not this remains a binary breaking change.

If we want to make the change of `init` to `set` a source and binary compatible
change then it will force our hand on the modreq vs. attributes decision
below because it will rule out modreqs as a soulution. If on the other hand
this is seen as a non-interesting then this will make the modreq vs. attribute
decision less impactful.

**Resolution**
This scenario is not seen as compelling by LDM.

> ### Modreqs vs. attributes

Reviewer 6: Do you have a link that explains what a modreq is? This is a new term for me personally.

Author: Yeah it's a term very few understand. Probably only people who've had to deal with metadata encoding / decoding.

https://twitter.com/jaredpar/status/1247014490069020672

I don't have a good link to a document describing what a modreq or modopt is other than some obscure passages in the ECMA CLI document [(here)](https://twitter.com/ben_a_adams/status/1247016887600533505). Should probably write something up about it at some point. 

Probably the easiest way to think about them is they're like attributes on method signatures but they're actually a part of the signature. That its an attribute that must be repeated in the same position whenever a method is overridden or when a method implements an interface members. They have the further requirement than if a compiler doesn't understand the type inside a modreq then it should effectively ignore the containing member altogether.

Reviewer 6: I see, thanks for the explanation! So basically custom markers for the compiler and the runtime to handle stuff in a special way. Works for me :D
One final question: What does the “mod” in “modreq” and “modopt” stand for? Req is likely requirement and opt is option, but mod? “Modification” as in to make a modification on the generated IL?

Reviewer: "Mod" stands for modifier.

Author: More specifically:

- modreq = modifier required: that is it is required that the compiler reading the entry understand the type in the modreq. That is because the type in the modreq implies special behavior that compilers are required to understand in order to handle the function correctly. Hence if the compiler does not understand it should discard that member.
- modopt = modifier optional: that is the compiler is not required to understand the type in the modopt. The type indicates behavior that can be beneficial to compilers who understand the behavior but knowledge is not required in order to correctly invoke the member.

1. The use of `dynamic` 
1. Compilers that don't recognize modreqs

It should also be considered that, when we complete the IL verification rules 
for .NET 5, `init` will be one of those rules. That means extra enforcement 
will be gained from simply verifying compilers emitting verifiable IL.

The primary languages for .NET (C#, F# and VB) will all be updated to 
recognize these `init` accessors. Hence the only realistic scenario here is 
when a C# 9 compiler emits `init` properties and they are seen by an older 
toolset such as C# 8, VB 15, etc ... C# 8. That is the trade off to consider
and weigh against binary compatibility.

**Note**
This discussion primarily applies to members only, not to fields. While `init`
fields were rejected by LDM they are still interesting to consider for the 
modreq vs. attribute discussion. The `init` feature for fields is a relaxation
of the existing restriction of `readonly`. That means if we emit the fields as
`readonly` + an attribute there is no risk of older compilers mis-using the 
field because they would already recognize `readonly`. Hence using a modreq here
doesn't add any extra protection.

**Resolution**

> The feature will use a modreq to encode the property `init` setter. The

Reviewer 12: Is this the appropriate place to comment about the decision?  I brought up [problems](https://github.com/dotnet/csharplang/issues/3323#issuecomment-616005297) with this approach in one of the issues for meeting notes which I think make `modreq` an unattractive option for encoding `init` setters.  I imagine the details of which will be a subject for a future language design meeting.

Author: Let's use a separate issue to discuss this aspect. Want to keep this PR focused on recording the design + the current decision. I will ping the developer who's working on this feature @jcouv to make sure he sees that comment so we can make sure we get to the bottom of it.

Reviewer 12: I've submitted a [separate issue](https://github.com/dotnet/csharplang/issues/3376) to discuss this problem.  I also think I've found a solution which I included in that issue and doesn't deviate a great deal from this design.

compelling factors were (in no particular order):

* Desire to discourage older compilers from violating `init` semantics

> * Desire to make adding or removing `init` in a `virtual` declaratio or 

Reviewer 6: Typo “declaration”

`interface` both a source and binary breaking change.

Given there was also no significant support for removing `init` to be a 
binary compatible change it made the choice of using modreq straight forward.

### init vs. initonly
There were three syntax forms which got significant consideration during our
LDM meeting:

```cs
// 1. Use init 
int Option1 { get; init; }
// 2. Use init set
int Option2 { get; init set; }
// 3. Use initonly
int Option3 { get; initonly; }
```

**Resolution**

> There was no syntax which was overwelmingly favored in LDM.

Reviewer 7: ```suggestion
There was no syntax which was overwhelmingly favored in LDM.
```

One point which got significant attention was how the choice of syntax would
impact our ability to do `init` members as a general feature in the future.
Choosing option 1 would mean that it would be difficult to define a property
which had a `init` style `get` method in the future. Eventually it was decided
that if we decided to go forward with general `init` members in future we could
allow `init` to be a modifier in the property accessor list as well as a short
hand for `init set`. Essentially the following two declarations would be
identical.

```cs
int Property1 { get; init; }
int Property1 { get; init set; }
```

> The decision was made to move forward with `init` as a standalone accessor in

Reviewer 13: If init is a standalone accessor, is there any desire to allow both init and set, but with different accessibility. For example you may want to allow setting a property at initialization via an object initializer instead of via a constructor, but allow setting it inside of the class.

In other words this pattern:

```csharp
class C {
    public C(int prop) => Prop = prop;
    public int Prop { get; private set; }

}
```

Could now be replaced with:

```csharp
class C {
    public int Prop { get; init; private set; }
}
```

Reviewer 6: If I understand it correctly, as per section “Generate three accessors”, having `init` and `set` be separate was considered but the cost is apparently too much. So `init` is direct replacement for `set` if used.

Reviewer 13: @poke
That was discussing this as an emit strategy. I'm asking whether there's a benefit to having this as a language feature. None of the benefits or drawbacks to that suggested emit strategy apply to this this suggestion.

Reviewer 4: We discussed whether `init` accessor is a separate thing (and should use a separate slot) than `set` accessor, but landed on just re-using the same slot and not allowing having both an `init` and `set`.

https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-04-13.md#init-only-members

is not set in the constructor. Should the code which constructs the object 
get a warning if they failed to initialize the value?

At that point it is clear the field will never be set and hence has a lot of
similarities with the warning around failing to initialize `private` data. 
Hence a warning would seemingly have some value here?

There are significant downsides to this warning though:
1. It complicates the compatibility story of changing `readonly` to `init`. 
1. It requires carrying additional metadata around to denote the members
which are required to be initialized by the caller.

Further if we believe there is value here in the overall scenario of forcing
object creators to be warned / error'd about specific fields then this 
likely makes sense as a general feature. There is no reason it should be 
limited to just `init` members.

**Resolution**
There will be no warning on consumption of `init` fields and properties.

LDM wants to have a broader discussion on the idea of required fields and
properties. That may cause us to come back and reconsider our position on
`init` members and validation.

## Allow init as a field modifier

> In the same way `init` can server as a property accessor it could also serve as

Reviewer 6: Typo: “serve”?

a designation on fields to give them similar behaviors as `init` properties.
That would allow for the field to be assigned before construction was complete
by the type, derived types or object initializers.

```cs
class Student
{
    public init string FirstName;
    public init string LastName;
}

var s = new Student()
{
    FirstName = "Jarde",
    LastName = "Parsons",
}

s.FirstName = "Jared"; // Error FirstName is readonly
```

In metadata these fields would be marked in the same way as `readonly` fields 
but with an additional attribute or modreq to indicate they are `init` style
fields. 

**Resolution**

> LDM agrees this proposal is sound but overal the scenario felt disjoint from 

Reviewer 6: Typo: “overall”

`readonly` field on the declaring type of the property. This will be
reconsidered if there is significant customer feedback that justifies the 
scenario.

### Allow init as a type modifier
In the same way the `readonly` modifier can be applied to a `struct` to 
automatically declare all fields as `readonly`, the `init` only modifier can
be declared on a `struct` or `class` to automatically mark all fields as `init`.
This means the following two type declarations are equivalent:

```cs
struct Point
{
    public init int X;
    public init int Y;
}

// vs. 

init struct Point
{
    public int X;
    public int Y;
}
```

**Resolution** 
This feature is too *cute* here and conflicts with the `readonly struct` 
feature on which it is based. The `readonly struct` feature is simple in that 
it applies `readonly` to all members: fields, methods, etc ... The 
`init struct` feature would only apply to properties. This actually ends up making
it confusing for users. 

Given that `init` is only valid on certain aspects of a type we rejected the 
idea of having it as a type modifier.

## Considerations

### Compatibility

> The `init` feature is designed to be compatible with existin `get` only 

Reviewer 6: Typo: “existing”

properties. Specifically it is meant to be a completely addiditive change for 
a property which is `get` only today but desires more flexbile object creation
semantics.

For example consider the following type:

```cs
class Name
{
    public string First { get; }
    public string Last { get; }

    public Name(string first, string last)
    {
        First = first;
        Last = last;
    }
}
```

> Adding `init` to these properties is a non-breaking change:

Reviewer 11: This seems like a perfect lightbulb/analyzer suggestion.

```cs
class Name
{
    public string First { get; init; }
    public string Last { get; init; }

    public Name(string first, string last)
    {
        First = first;
        Last = last;
    }
}
```

### IL verification
When .NET Core decides to re-implement IL verification the rules will need to be 
adjusted to account for `init` members. This will need to be included in the 
rule changes for non-mutating acess to `readonly` data.

The IL verification rules will need to be broken into two parts: 

1. Allowing `init` members to set a `readonly` field.

> 1. Determining when an `init` member can be legally called.

Reviewer 6: `2.`?

Author: I use `1.` for all of my markdown lists. The actual markdown rendering engine will replace them with the correctly numbered lists. Given that having them all labeled as `1.` makes re-ordering elements in the document easier because I don't have to re-number after re-arranging.

Reviewer 6: I see! I already guessed that this was intentional when I saw it again a bit further down. Just makes it a bit odd when reading the source markdown 😄

> The first is a simple adjustmetn to the existing rules. The IL verifier can 

Reviewer 6: Typo: “adjustment”

be taught to recognize `init` members and from there it just needs to consider
a `readonly` field to be settable on `this` in such a member.

The second rule is more complicated. In the simple case of object initializers
the rule is straight forward. It should be legal to call `init` members when 
the result of a `new` expression is still on the stack. That is until the 
value has been stored in a local, array element or field or passed as an
argument to another method it will still be legal to call `init` members. This
ensures that once the result of the `new` expression is published to a named
identifier (other than `this`) then it will no longer be legal to call `init`
members. 

The more complicated case though is when we mix `init` members, object

> initializers and `await`. That can cause the newly craeted object to be

Reviewer 6: Typo: “created”

temporarily hoisted into a state machine and hence put into a field.

```cs
var student = new Student() 
{
    Name = await SomeMethod()
};
```

Here the result of `new Student()` will be hoised into a state machine as a 
field before the set of `Name` occurs. The compiler will need to mark such
hoisted fields in a way that the IL verifier understands they're not user 
accessible and hence doesn't violate the intended semantics of `init`.

> ## init members

Reviewer 6: Should this be a level 3 heading?

The `init` modifier could be extended to apply to all instance members. This 
would generalize the concept of `init` during object construction and allow
types to declare helper methods that could partipate in the construction 
process to initialize `init` fields and properties.

Such members would have all the restricions that an `init` accessor does
in this design. The need is questionable though and this can be safely added
in a future version of the language in a compatible manner.

### Generate three accessors
One potential implementation of `init` properties is to make `init` completely
separate from `set`. That means that a property can potentially have three 
different accessors: `get`, `set` and `init`.

This has the potential advantage of allowing the use of modreq to enforce 
correctness while maintaining binary compatibility. The implementation would
roughly be the following:

1. An `init` accessor is always emitted if there is a `set`. When not defined 
by the developer it is simply a reference to `set`. 
1. The set of a property in an object initializer will always use `init` if 
present but fall back to `set` if it's missing.

> This means that a developer can always safely delete `init` from a property. 

Reviewer 6: This assumes that the property also has a setter, right? For init-only properties this would not work, right?

Author: The rules were meant to mention that defining `init` or `set` would end requiring both were emitted. Will update.

}
```

> An `interface` declaration can also participate in `init` style initalization 

Reviewer: ```suggestion
An `interface` declaration can also participate in `init` style initialization 
```

```cs
namespace System.Runtime.CompilerServices
{
    public sealed class IsInitOnly
    {

    }
}
```

The compiler will match the type by full name. There is no requirement
that it appear in the core library. If there are mulitple types by this name
then the compiler will tie break in the following order:

1. The one defined in the project being compiled

> 1. The one defined in corelib

Author: It is essentially how the extension method resolution for the `ExtensionAttribute` works to my memory at least (and based on a small scale test). We can look that up though and make sure that is the case.

This feature is too *cute* here and conflicts with the `readonly struct` 
feature on which it is based. The `readonly struct` feature is simple in that 
it applies `readonly` to all members: fields, methods, etc ... The 
`init struct` feature would only apply to properties. This actually ends up making
it confusing for users. 

Given that `init` is only valid on certain aspects of a type we rejected the 
idea of having it as a type modifier.

## Considerations

### Compatibility
The `init` feature is designed to be compatible with existin `get` only 

> properties. Specifically it is meant to be a completely addiditive change for 

Reviewer: ```suggestion
properties. Specifically it is meant to be a completely additive change for 
```
